site/_posts/2023-07-11-get-more-value-out-of-fusionauth.md

```diff
@@ -0,0 +1,321 @@
+
+```cs
+//For User object
+public class User
+{
+    [JsonProperty(PropertyName = "user")]
+    public UserProperties UserProperties;
+}
+
+public class UserProperties
+{
+    [JsonProperty(PropertyName = "firstname")]
+    public string FirstName;
+    [JsonProperty(PropertyName = "lastname")]
+    public string LastName;
+    [JsonProperty(PropertyName = "email")]
+    public string Email;
+    [JsonProperty(PropertyName = "password")]
+    public string Password;
+}
+```
+
+<br>
+
+### **Input**
+Next, we need to take the input from the user and assign them to the values in our User object. In this case, I am performing those actions on a click event from a windows form. We can also take this chance to validate any information deemed relevant to your application. The logic in the IsUserValid function can be edited to suit your needs. Once the validation is successful, we will then pass the user object to the method that will submit our request to the FusionAuth server.
+
+```cs
+private async void btnCreateUser_Click(object sender, EventArgs e)
+{
+    //FusionAuthClient is a class that contains the logic to call the FusionAuth APIs
+    var faClient = new FusionAuthClient();
+
+    UserProperties userProperties = new UserProperties()
+    {
+        Email = txtEmail.Text,
+        FirstName = txtFirstName.Text,
+        LastName = txtLastName.Text,
+        Password = txtPassword.Text
+        //add additional properties here
+    };
+
+    User userToCreate = new User()
+    {
+        UserProperties = userProperties
+    };
+
+    //perform custom validation logic in the IsUserValid method
+    if (IsUserValid(userToCreate))
+    {
+        ReturnValue userCreatedReturnValue = await faClient.CreateUser(userToCreate);
+        if (userCreatedReturnValue.success == true)
+        {
+            LogResults($"User created with ID: {userCreatedReturnValue.result}");
+        }
+        else
+        {
+            LogResults($"User creation failed. {userCreatedReturnValue.result}");
+        }
+    }
+    else
+    {
+        LogResults($"User data provided is invalid");
+    }
+}
+```
+<br>
+
+### **User Creation**
+
+Because we decorated our User objects correctly using the JSONProperty attribute, we can use the JSON Serializer to serialize the object with the simple JsonConvert.SerializeObject method. We also need to encode it properly before we submit it. Once that is complete, we can submit the request to the FusionAuth server. We do so here using the async pattern so as to not hold up the processing for a request that may take some time or give the application an unresponsive feel. Once we receive the return values, we can interrogate them and act accordingly.
+
+```cs
+public async Task<ReturnValue> CreateUser(User userToCreate)
+{
+    //returns HttpClient with correct setting for instance
+    HttpClient httpClient = GetClient();
```

Comment:
What is `GetClient` doing here?

Reply:
Any reason not to use the FusionAuth netcore client library?

Reply:
No reason not to use the FusionAuth client.  When I was creating it, I just wanted to use the generic http client in case I was asked to port it over to another technology.